File: Ndapi/LOV.cpp
#include "stdafx.h"
#include "LOV.h"

namespace Ndapi
{
	LOV::LOV() : NdapiObject()
	{
	}

	LOV::LOV(d2fob* alert) : NdapiObject(alert)
	{
	}

	LOV::LOV(FormModule^ form, String^ name)
	{
		Create(name, form, D2FFO_LOV);
	}

	bool LOV::AutomaticColumnWidth::get()
	{
		return GetBooleanProperty(D2FP_AUTO_COL_WID);
	}

	void LOV::AutomaticColumnWidth::set(bool value)
	{
		SetBooleanProperty(D2FP_AUTO_COL_WID, value);
	}

	bool LOV::AutomaticDisplay::get()
	{
		return GetBooleanProperty(D2FP_AUTO_DISP);
	}

	void LOV::AutomaticDisplay::set(bool value)
	{
		SetBooleanProperty(D2FP_AUTO_DISP, value);
	}

	bool LOV::AutomaticPosition::get()
	{
		return GetBooleanProperty(D2FP_AUTO_POS);
	}

	void LOV::AutomaticPosition::set(bool value)
	{
		SetBooleanProperty(D2FP_AUTO_POS, value);
	}

	bool LOV::AutomaticRefresh::get()
	{
		return GetBooleanProperty(D2FP_AUTO_RFRSH);
	}

	void LOV::AutomaticRefresh::set(bool value)
	{
		SetBooleanProperty(D2FP_AUTO_RFRSH, value);
	}

	bool LOV::AutomaticSkip::get()
	{
		return GetBooleanProperty(D2FP_AUTO_SKP);
	}

	void LOV::AutomaticSkip::set(bool value)
	{
		SetBooleanProperty(D2FP_AUTO_SKP, value);
	}

	bool LOV::AutomaticSelect::get()
	{
		return GetBooleanProperty(D2FP_AUTO_SLCT);
	}

	void LOV::AutomaticSelect::set(bool value)
	{
		SetBooleanProperty(D2FP_AUTO_SLCT, value);
	}

	String^ LOV::BackgroundColor::get()
	{
		return GetStringProperty(D2FP_BACK_COLOR);
	}

	void LOV::BackgroundColor::set(String^ value)
	{
		SetStringProperty(D2FP_BACK_COLOR, value);
	}

	NdapiObject^ LOV::ColumnMappingProperties::get()
	{
		return GetObjectProperty<NdapiObject^>(D2FP_COL_MAP);
	}

	String^ LOV::Comment::get()
	{
		return GetStringProperty(D2FP_COMMENT);
	}

	void LOV::Comment::set(String^ value)
	{
		SetStringProperty(D2FP_COMMENT, value);
	}

	String^ LOV::FillPattern::get()
	{
		return GetStringProperty(D2FP_FILL_PAT);
	}

	void LOV::FillPattern::set(String^ value)
	{
		SetStringProperty(D2FP_FILL_PAT, value);
	}

	bool LOV::FilterBeforeDisplay::get()
	{
		return GetBooleanProperty(D2FP_FLTR_BEFORE_DISP);
	}

	void LOV::FilterBeforeDisplay::set(bool value)
	{
		SetBooleanProperty(D2FP_FLTR_BEFORE_DISP, value);
	}

	String^ LOV::FontName::get()
	{
		return GetStringProperty(D2FP_FONT_NAM);
	}

	void LOV::FontName::set(String^ value)
	{
		SetStringProperty(D2FP_FONT_NAM, value);
	}

	long LOV::FontSize::get()
	{
		return GetNumberProperty(D2FP_FONT_SIZ);
	}

	void LOV::FontSize::set(long value)
	{
		SetNumberProperty(D2FP_FONT_SIZ, value);
	}

	Ndapi::FontSpacing LOV::FontSpacing::get()
	{
		return GetNumberProperty<Ndapi::FontSpacing>(D2FP_FONT_SPCING);
	}

	void LOV::FontSpacing::set(Ndapi::FontSpacing value)
	{
		SetNumberProperty(D2FP_FONT_SPCING, value);
	}

	Ndapi::FontStyle LOV::FontStyle::get()
	{
		return GetNumberProperty<Ndapi::FontStyle>(D2FP_FONT_STY);
	}

	void LOV::FontStyle::set(Ndapi::FontStyle value)
	{
		SetNumberProperty(D2FP_FONT_STY, value);
	}

	Ndapi::FontWeight LOV::FontWeight::get()
	{
		return GetNumberProperty<Ndapi::FontWeight>(D2FP_FONT_WGHT);
	}

	void LOV::FontWeight::set(Ndapi::FontWeight value)
	{
		SetNumberProperty(D2FP_FONT_WGHT, value);
	}

	String^ LOV::ForegroundColor::get()
	{
		return GetStringProperty(D2FP_FORE_COLOR);
	}

	void LOV::ForegroundColor::set(String^ value)
	{
		SetStringProperty(D2FP_FORE_COLOR, value);
	}

	long LOV::Height::get()
	{
		return GetNumberProperty(D2FP_HEIGHT);
	}

	void LOV::Height::set(long value)
	{
		SetNumberProperty(D2FP_HEIGHT, value);
	}

	Ndapi::LanguageDirection LOV::LanguageDirection::get()
	{
		return GetNumberProperty<Ndapi::LanguageDirection>(D2FP_LANG_DIR);
	}

	void LOV::LanguageDirection::set(Ndapi::LanguageDirection value)
	{
		SetNumberProperty(D2FP_LANG_DIR, value);
	}

	Ndapi::ListType LOV::ListType::get()
	{
		return GetNumberProperty<Ndapi::ListType>(D2FP_LST_TYP);
	}

	void LOV::ListType::set(Ndapi::ListType value)
	{
		SetNumberProperty(D2FP_LST_TYP, value);
	}

	String^ LOV::Name::get()
	{
		return GetStringProperty(D2FP_NAME);
	}

	void LOV::Name::set(String^ value)
	{
		SetStringProperty(D2FP_NAME, value);
	}

	String^ LOV::OldLOVText::get()
	{
		return GetStringProperty(D2FP_OLD_LOV_TXT);
	}

	void LOV::OldLOVText::set(String^ value)
	{
		SetStringProperty(D2FP_OLD_LOV_TXT, value);
	}

	Ndapi::RecordGroup^ LOV::RecordGroup::get()
	{
		return GetObjectProperty<Ndapi::RecordGroup^>(D2FP_REC_GRP_OBJ);
	}

	void LOV::RecordGroup::set(Ndapi::RecordGroup^ value)
	{
		SetObjectProperty(D2FP_REC_GRP_OBJ, value);
	}

	String^ LOV::Title::get()
	{
		return GetStringProperty(D2FP_TITLE);
	}

	void LOV::Title::set(String^ value)
	{
		SetStringProperty(D2FP_TITLE, value);
	}

	String^ LOV::VisualAttributeGroup::get()
	{
		return GetStringProperty(D2FP_VAT_NAM);
	}

	void LOV::VisualAttributeGroup::set(String^ value)
	{
		SetStringProperty(D2FP_VAT_NAM, value);
	}

	bool LOV::WhiteOnBlack::get()
	{
		return GetBooleanProperty(D2FP_WHITE_ON_BLACK);
	}

	void LOV::WhiteOnBlack::set(bool value)
	{
		SetBooleanProperty(D2FP_WHITE_ON_BLACK, value);
	}

	long LOV::Width::get()
	{
		return GetNumberProperty(D2FP_WIDTH);
	}

	void LOV::Width::set(long value)
	{
		SetNumberProperty(D2FP_WIDTH, value);
	}

	long LOV::XPosition::get()
	{
		return GetNumberProperty(D2FP_X_POS);
	}

	void LOV::XPosition::set(long value)
	{
		SetNumberProperty(D2FP_X_POS, value);
	}

	long LOV::YPosition::get()
	{
		return GetNumberProperty(D2FP_Y_POS);
	}

	void LOV::YPosition::set(long value)
	{
		SetNumberProperty(D2FP_Y_POS, value);
	}
}

File: Ndapi/LOV.cs
using Ndapi.Core;
using Ndapi.Core.Handles;
using Ndapi.Enums;

namespace Ndapi;

/// <summary>
/// Represents a list of values.
/// </summary>
public partial class Lov : NdapiObject<Lov>
{
    /// <summary>
    /// Creates a list of values.
    /// </summary>
    /// <param name="module">List of values owner.</param>
    /// <param name="name">List of values name.</param>
    public Lov(FormModule module, string name) : base(name, ObjectType.Lov, module)
    {
    }

    /// <summary>
    /// Creates a list of values.
    /// </summary>
    /// <param name="group">List of values owner.</param>
    /// <param name="name">List of values name.</param>
    public Lov(ObjectGroup group, string name) : base(name, ObjectType.Lov, group)
    {
    }

    internal Lov(ObjectSafeHandle handle) : base(handle, ObjectType.Lov)
    {
    }

    /// <summary>
    /// Gets or sets whether runtime should adjust the width of columns automatically.
    /// </summary>
    [Property(NdapiConstant.D2FP_AUTO_COL_WID)]
    public partial bool IsAutomaticColumnWidth { get; set; }

    /// <summary>
    /// Gets or sets whether runtime should display the LOV automatically when the user or
    /// the application navigates into a text item to which the LOV is attached.
    /// </summary>
    [Property(NdapiConstant.D2FP_AUTO_DISP)]
    public partial bool IsAutomaticDisplay { get; set; }

    /// <summary>
    /// Gets or sets whether runtime should position the LOV automatically near the field from which it was invoked. 
    /// </summary>
    [Property(NdapiConstant.D2FP_AUTO_POS)]
    public partial bool IsAutomaticPosition { get; set; }

    /// <summary>
    /// Gets or sets whether runtime re-executes the query to populate an LOV that is based on a query record group.
    /// </summary>
    [Property(NdapiConstant.D2FP_AUTO_RFRSH)]
    public partial bool IsAutomaticRefresh { get; set; }

    /// <summary>
    /// Gets or sets whether runtime moves the cursor to the next navigable item when the user makes 
    /// a selection from an LOV to a text item. 
    /// </summary>
    [Property(NdapiConstant.D2FP_AUTO_SKP)]
    public partial bool IsAutomaticSkip { get; set; }

    /// <summary>
    /// Gets or sets whether runtime should select automatically the row when the result list is reduced to a single choice
    /// (by using auto-reduction or searching). If true, the LOV is dismissed automatically and column values from the 
    /// single row are assigned to their corresponding return items. If false, the LOV remains displayed, giving the user 
    /// the option to explicitly select the remaining choice or dismiss the LOV.
    /// </summary>
    [Property(NdapiConstant.D2FP_AUTO_SLCT)]
    public partial bool IsAutomaticSelect { get; set; }

    /// <summary>
    /// Gets or sets the background color.
    /// </summary>
    [Property(NdapiConstant.D2FP_BACK_COLOR)]
    public partial string BackgroundColor { get; set; }

    /// <summary>
    /// Gets or sets the comment.
    /// </summary>
    [Property(NdapiConstant.D2FP_COMMENT)]
    public partial string Comment { get; set; }

    /// <summary>
    /// Gets or sets the pattern to be used for the fill region.
    /// Patterns are rendered in the two colors specified by <see cref="BackgroundColor"/> and <see cref="ForegroundColor"/>.
    /// </summary>
    [Property(NdapiConstant.D2FP_FILL_PAT)]
    public partial string FillPattern { get; set; }

    /// <summary>
    /// Gets or sets whether runtime allows the user to enter a query criteria before displaying the query results.
    /// </summary>
    [Property(NdapiConstant.D2FP_FLTR_BEFORE_DISP)]
    public partial bool FilterBeforeDisplay { get; set; }

    /// <summary>
    /// Gets or sets the font name.
    /// </summary>
    [Property(NdapiConstant.D2FP_FONT_NAM)]
    public partial string FontName { get; set; }

    /// <summary>
    /// Gets or sets the font size.
    /// </summary>
    [Property(NdapiConstant.D2FP_FONT_SIZ)]
    public partial int FontSize { get; set; }

    /// <summary>
    /// Gets or sets the font spacing.
    /// </summary>
    [Property(NdapiConstant.D2FP_FONT_SPCING)]
    public partial FontSpacing FontSpacing { get; set; }

    /// <summary>
    /// Gets or sets the font style.
    /// </summary>
    [Property(NdapiConstant.D2FP_FONT_STY)]
    public partial FontStyle FontStyle { get; set; }

    /// <summary>
    /// Gets or sets the font weight.
    /// </summary>
    [Property(NdapiConstant.D2FP_FONT_WGHT)]
    public partial FontWeight FontWeight { get; set; }

    /// <summary>
    /// Gets or sets the foreground color.
    /// </summary>
    [Property(NdapiConstant.D2FP_FORE_COLOR)]
    public partial string ForegroundColor { get; set; }

    /// <summary>
    /// Gets or sets the height.
    /// </summary>
    [Property(NdapiConstant.D2FP_HEIGHT)]
    public partial int Height { get; set; }

    /// <summary>
    /// Gets or sets the language direction.
    /// </summary>
    [Property(NdapiConstant.D2FP_LANG_DIR)]
    public partial LanguageDirection LanguageDirection { get; set; }

    /// <summary>
    /// Gets or sets the list type.
    /// </summary>
    [Property(NdapiConstant.D2FP_LST_TYP)]
    public partial ListType ListType { get; set; }

    /// <summary>
    /// Gets or sets the LOV text (included only for compatibility).
    /// </summary>
    [Property(NdapiConstant.D2FP_OLD_LOV_TXT)]
    public partial string OldLovText { get; set; }

    /// <summary>
    /// Gets or sets the record group associated with the LOV.
    /// </summary>
    [Property(NdapiConstant.D2FP_REC_GRP_OBJ)]
    public partial RecordGroup RecordGroup { get; set; }

    /// <summary>
    /// Gets or sets the LOV title.
    /// </summary>
    [Property(NdapiConstant.D2FP_TITLE)]
    public partial string Title { get; set; }

    /// <summary>
    /// Gets or sets the visual attribute name.
    /// </summary>
    [Property(NdapiConstant.D2FP_VAT_NAM)]
    public partial string VisualAttributeName { get; set; }

    /// <summary>
    /// Gets or sets the white-on-black property.
    /// </summary>
    [Property(NdapiConstant.D2FP_WHITE_ON_BLACK)]
    public partial bool IsWhiteOnBlack { get; set; }

    /// <summary>
    /// Gets or sets the width.
    /// </summary>
    [Property(NdapiConstant.D2FP_WIDTH)]
    public partial int Width { get; set; }

    /// <summary>
    /// Gets or sets the X position measured at the upper left corner.
    /// </summary>
    [Property(NdapiConstant.D2FP_X_POS)]
    public partial int XPosition { get; set; }

    /// <summary>
    /// Gets or sets the Y position measured at the upper left corner.
    /// </summary>
    [Property(NdapiConstant.D2FP_Y_POS)]
    public partial int YPosition { get; set; }

    /// <summary>
    /// Gets all the LOV column mappings.
    /// </summary>
    [Property(NdapiConstant.D2FP_COL_MAP)]
    public partial NdapiObjectList<ColumnMapping> ColumnMappings { get; }

    /// <summary>
    /// Creates a column mapping.
    /// </summary>
    /// <param name="name">Name of the column mapping.</param>
    /// <returns>The child object.</returns>
    public ColumnMapping CreateColumnMapping(string name) => new(this, name);
}
